### PR TITLE
REL: 2.3.1

### PR DESCRIPTION
**Pull request by effigies**
Built a changelog, and think it's about time for a release. Any objections? Predominantly bug fixes, so I'm calling it 2.3.1 unless Python 3.7 support or `nib-diff` justifies a new minor version.

Open issues/PRs that could go in quickly
--------
* [x] #661 [PR] (Submitter: @yarikoptic; Reviewer: @chrispycheng)
* [ ] #652 [PR] (Submitter: @mattcieslak)
* [ ] #644 [PR] (Submitter: @effigies; Reviewer: @matthew-brett)
* [x] #619 [BUG] (Submitter: @larsoner)

Pre-release checklist
--------
* [x] Review the open list of [nibabel issues](https://github.com/nipy/nibabel/issues). Check whether there are outstanding issues that can be closed, and whether there are any issues that should delay the release. Label them!
* [x] Review and update the release notes. Review and update the [Changelog file](https://github.com/nipy/nibabel/blob/master/Changelog).
* [x] Look at [`doc/source/index.rst`](https://github.com/nipy/nibabel/blob/master/doc/source/index.rst) and add any authors not yet acknowledged.
* [x] Use the opportunity to update the [`.mailmap file`](https://github.com/nipy/nibabel/blob/master/.mailmap) if there are any duplicate authors listed from `git shortlog -nse`.
* [x] Check the copyright year in [doc/source/conf.py](https://github.com/nipy/nibabel/blob/master/doc/source/conf.py)
* [x] Refresh the [README.rst](https://github.com/nipy/nibabel/blob/master/README.rst) text from the `LONG_DESCRIPTION` in [`info.py`](https://github.com/nipy/nibabel/blob/master/nibabel/info.py) by running `make refresh-readme`.
* [x] Check the dependencies listed in [`nibabel/info.py`](https://github.com/nipy/nibabel/blob/master/nibabel/info.py) (e.g. `NUMPY_MIN_VERSION`) and in [`doc/source/installation.rst`](https://github.com/nipy/nibabel/blob/master/doc/source/installation.rst) and in [`requirements.txt`](https://github.com/nipy/nibabel/blob/master/requirements.txt) and [`.travis.yml`](https://github.com/nipy/nibabel/blob/master/.travis.yml). They should at least match. Do they still hold? Make sure nibabel on travis is testing the minimum dependencies specifically.
* [x] Make sure all tests pass (from the nibabel root directory): `nosetests --with-doctest nibabel`
* [x] Edit [`nibabel/info.py`](https://github.com/nipy/nibabel/blob/master/nibabel/info.py) to set `_version_extra` to `''`; commit
* [x] Build bots
  * [x] [nibabel-py2.7-osx-10.10](https://nipy.bic.berkeley.edu/builders/nibabel-py2.7-osx-10.10/builds/205/steps/shell_6/logs/stdio)
  * [x] [nibabel-bdist32-27](http://nipy.bic.berkeley.edu/builders/nibabel-bdist32-27/builds/227/steps/shell_8/logs/stdio)
  * [x] [nibabel-bdist32-34](https://nipy.bic.berkeley.edu/builders/nibabel-bdist32-34/builds/214/steps/shell_8/logs/stdio)
  * [x] [nibabel-bdist32-35](https://nipy.bic.berkeley.edu/builders/nibabel-bdist32-35/builds/136/steps/shell_8/logs/stdio)
  * [ ] nibabel-bdist64-27 [buildslave offline]

Adapted from http://nipy.org/nibabel/devel/make_release.html#release-checklist

### CODE REVIEW

**Comment by matthew-brett** [comment]
A release is a good idea - thanks for the proposal.  Sorry, I am very busy at the moment with term starting.

**Comment by matthew-brett** [comment]
Sure, sure, please go ahead.

**Comment by effigies** [comment]
@yarikoptic Any Debian-related tests you want to run before release? Also interested in trying to get #661 in? (I haven't reviewed. Let me know if you want another pair of eyes.)

**Comment by yarikoptic** [comment]
Review/feedback on #661 would be very much appreciated, and inclusion in the release as well!
Meanwhile I will run a test build on debians using current master

**Comment by yarikoptic** [comment]
or better -- I will use this PR to base the debian pkg on
Cheers

**Comment by effigies** [comment]
Yes, please use this PR. I'm merging into the `maint/2.3.x` branch. Though the diff is ~2 lines of version code.

**Comment by yarikoptic** [comment]
btw, for the release tag, please make sure it is annotated, `2.3.0` wasn't 
```shell
$> git describe --tags --match 2* pr-667        
2.3.0-121-ge493b9dd

$> git describe  --match 2* pr-667    
2.2.1-383-ge493b9dd
```

File: Changelog [hunk 1]
```diff
@@ -24,6 +24,38 @@ Gerhard (SG) and Eric Larson (EL).
 
 References like "pr/298" refer to github pull request numbers.
 
+Upcoming release
+================
+
+New features
+------------
+* ``nib-diff`` command line tool for comparing image files (pr/617)
+  (Chris Cheng, reviewed by Yaroslav Halchenko, Pradeep Raamana and CM)
```
Comment (yarikoptic) [inline Changelog]: I love the appearance of my name, but you could add me and Chris with abbreviations (YOH, CC?) as "Other contributors" in the section above where you introduce them in general.  I think we will keep appearing here from time to time ;)

**Comment by yarikoptic** [comment]
apparently we with @chrispycheng have added a versioned dependency on hypothesis, e.g. now somewhat older debian/ubuntus fail with
```
======================================================================
ERROR: nibabel.tests.test_diff.test_diff_values_int
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/build/nibabel-2.3.0+git118-ge493b9dd/nibabel/tests/test_diff.py", line 22, in test_diff_values_int
    def test_diff_values_int(data):
  File "/usr/lib/python2.7/dist-packages/hypothesis/core.py", line 520, in wrapped_test
    print_example=True, is_final=True
  File "/usr/lib/python2.7/dist-packages/hypothesis/executors.py", line 58, in default_new_style_executor
    return function(data)
  File "/usr/lib/python2.7/dist-packages/hypothesis/core.py", line 110, in run
    return test(*args, **kwargs)
  File "/build/nibabel-2.3.0+git118-ge493b9dd/nibabel/tests/test_diff.py", line 23, in test_diff_values_int
    x = data.draw(st.integers(), label='x')
TypeError: draw() got an unexpected keyword argument 'label'
```
that additional argument was added (according to git log exploration) in 3.7.0 of hypothesis, which fits the version information of hypothesis on those systems which fail.  Although `label` is optional, I think the "easiest" course of action will be for me to provide hypothesis backport for those systems and nibabel specify minimal required version of it

**Comment by effigies** [comment]
So to be clear, on my end I just need to update to `hypothesis>=3.7` in `dev-requirements.txt`?

**Comment by yarikoptic** [comment]
I think so... Bet let me also first do those backports and see if all is kosher

**Comment by yarikoptic** [comment]
ok, now that no hypothesis etc, I will give another round of builds!

**Comment by yarikoptic** [comment]
oh... so the release would be coming from rel/2.3.1 which is diverged from master:
```
$> git diff origin/master...origin/rel/2.3.1 --stat
 .mailmap             |  1 +
 Changelog            | 34 +++++++++++++++++++++++++++++++++-
 dev-requirements.txt |  2 +-
 nibabel/info.py      |  8 ++++----
```
will you @effigies remerge master into release branch or the plan is different and you would like to keep some stuff in master away from the release? (although it seems just diff related:
```shell
$> git diff origin/rel/2.3.1...origin/master --stat
 .travis.yml                         |   4 +-
 appveyor.yml                        |   2 +-
 dev-requirements.txt                |   1 -
 nibabel/cmdline/diff.py             | 240 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++------------
 nibabel/cmdline/tests/test_utils.py |  50 ++++++++++++--
 nibabel/tests/test_diff.py          | 102 +++++++++++++++--------------
 nibabel/tests/test_scripts.py       |   8 ++-
```
meanwhile I will locally merge master into rel/2.3.1 for testing

**Comment by effigies** [comment]
Merged master in. Will update the changelog when I get another minute.

**Comment by effigies** [comment]
@yarikoptic Nearly ready. Are your tests all green? (There was a very minor pull this morning, but if you want to do another round, fine with me.)

**Comment by pauldmccarthy** [comment]
Hey @effigies, when are you intending to release? I might have just found an issue in the mess that is `keep_file_open`+`indexed_gzip.drop_handles`. Am aiming to resolve it and get a PR up within the next hour or so...

**Comment by effigies** [comment]
Plan was today, but if there's a bug, I can wait. Can you start a PR with a test that fails, even before fixing it? That will make reviewing simpler.

**Comment by pauldmccarthy** [comment]
Sure, I'll do so ASAP. I think the existing `ArrayProxy` are invalid, so I'll fix them  as a first step...

**Comment by pauldmccarthy** [comment]
Started a PR at #679 ...

**Comment by effigies** [comment]
@yarikoptic Final tests okay? (I can't remember what you called them...)

**Pull request merged**